sdk/metric/integrator/simple/simple.go

```diff
@@ -68,97 +84,209 @@ type (
 	}
-	b.batch.values[key] = batchValue{
-		aggregator: agg,
-		labels:     accumulation.Labels(),
-		resource:   accumulation.Resource(),
+
+	// Advance the update sequence number:
+	sameRound := b.state.finishedCollection == value.updated
+	value.updated = b.state.finishedCollection
+
+	// An existing record will be found when:
+	// (a) stateful aggregation is required for an exporter
+	// (b) multiple accumulators (SDKs) are being used.
+	// Another accumulator must have produced this.
+
+	if !sameRound {
+		// This is the first time through in a new round.
+		value.current = agg
+		return nil
 	}
-	return nil
+	if desc.MetricKind().Asynchronous() {
+		// The last value across multiple accumulators is taken.
+		value.current = agg
+		return nil
+	}
+	if value.delta == nil {
+		// Merging values: may need to allocate the delta aggregator.
+		b.AggregationSelector.AggregatorFor(desc, &value.delta)
+	}
+	if value.current != value.delta {
+		// Merging two values, first copy the singleton.
+		err := value.current.SynchronizedCopy(value.delta, desc)
+		if err != nil {
+			return err
+		}
+		value.current = value.delta
+	}
+	return value.delta.Merge(agg, desc)
 }
 
+// CheckpointSet returns the associated CheckpointSet.  Use the
+// CheckpointSet Locker interface to synchronize access to this
+// object.  The CheckpointSet.ForEach() method cannot be called
+// concurrently with Process().
 func (b *Integrator) CheckpointSet() export.CheckpointSet {
-	return &b.batch
+	return &b.state
 }
 
+// StartCollection signals to the Integrator one or more Accumulators
+// will begin calling Process() calls during collection.
 func (b *Integrator) StartCollection() {
 	if b.startedCollection != 0 {
 		b.intervalStart = b.intervalEnd
 	}
 	b.startedCollection++
-	if !b.stateful {
-		b.batch.values = map[batchKey]batchValue{}
-	}
 }
 
+// FinishCollection signals to the Integrator that a complete round of
+// collection is finished and that ForEach will be called to access
+// the CheckpointSet.
 func (b *Integrator) FinishCollection() error {
-	b.finishedCollection++
 	b.intervalEnd = time.Now()
-	if b.startedCollection != b.finishedCollection {
+	if b.startedCollection != b.finishedCollection+1 {
 		return ErrInconsistentState
 	}
+	defer func() { b.finishedCollection++ }()
+
+	for key, value := range b.values {
+		mkind := key.descriptor.MetricKind()
+
+		if !value.stateful {
+			if value.updated != b.finishedCollection {
+				delete(b.values, key)
+			}
+			continue
+		}
+
+		var err error
+		if mkind.PrecomputedSum() {
+			// We need to compute a delta.  We have the prior cumulative value.
+			if subt, ok := value.current.(export.Subtractor); ok {
+				err = subt.Subtract(value.cumulative, value.delta, key.descriptor)
+
+				if err == nil {
+					err = value.current.SynchronizedCopy(value.cumulative, key.descriptor)
+				}
+			} else {
+				err = aggregation.ErrNoSubtraction
+			}
+		} else {
+			err = value.cumulative.Merge(value.current, key.descriptor)
```

Comment:
Above we explicitly do a subtraction. Here we do a `Merge` which iiuc because of the conditions to get to this point, will always have the effect of performing an addition?

If that's correct, perhaps we could make the addition here explicit so that's its consistent with the above code and really clear that we are just converting back and forth between cumulatives & deltas? (I might be totally wrong)

Also maybe consistent comments would help:

Above: `// delta value = current cumulative value - previous cumulative value`
Here: `// cumulative value = previous cumulative value + current delta value`

Reply:
I've added your recommended comments.

We could consider renaming the `Merge()` function to `Add()`, but there are cases where that may create confusion. There is (maybe) a fine distinction here: Aggregator's aren't required to implement Subtract, they should only do so when they deal in "Additive" measurements.  As it stands, "Merge" means Add for the Adding instruments and it means Combine for the Grouping instruments.

Reply:
Subtract is only used in the case where an instrument and exporter use PrecomputedSum **AND** is stateful because of a delta exporter. Subtract will be needed by default, only for a (dog)statsd exporter with (UpDown)SumObserver instruments.

Merge is the ordinary case, handling all the other possibilities including both Adding and Grouping instruments. 

I added support for Sum.Subtract in this PR.  If we wanted to compute a distribution of SumObserver values (as opposed to a simple sum), we'd choose a stateful Histogram aggregator which can also support Subtract (however I did not add it in this PR).